### PR TITLE
Mutable users vs authorizedKeysFiles

### PR DESCRIPTION
**Pull request by ip1981**
No mutable SSH keys by default for immutable users
    
There are security concerns: someone could upload a new SSH key,
and we will never know. This still can be extended or overridden.
    
The %h placeholder is added to fit both sshd_config and pam_ssh_agent_auth.
Also synchronize keys with openssh.

### CODE REVIEW

**Review by LnL7 — COMMENTED** [review]
Changes behaviour and might break existing setups, this definitively needs a changelog entry. And maybe we should keep the current behaviour for `stateVersion < 18.09` to avoid potentially loosing access after upgrading.

File: nixos/modules/security/pam.nix [hunk 1]
```diff
@@ -287,7 +287,7 @@ let
           ${optionalString cfg.logFailures
               "auth required pam_tally.so"}
           ${optionalString (config.security.pam.enableSSHAgentAuth && cfg.sshAgentAuth)
-              "auth sufficient ${pkgs.pam_ssh_agent_auth}/libexec/pam_ssh_agent_auth.so file=~/.ssh/authorized_keys:~/.ssh/authorized_keys2:/etc/ssh/authorized_keys.d/%u"}
+              "auth sufficient ${pkgs.pam_ssh_agent_auth}/libexec/pam_ssh_agent_auth.so file=${concatStringsSet ":" config.services.openssh.authorizedKeysFiles}"}
```
Comment (adisbladis) [inline nixos/modules/security/pam.nix]: `concatStringsSet` -> `concatStringsSep`

**Review by tomberek — CHANGES_REQUESTED** [review]
This may be surprising, so it can use a bit of documentation under mutableUsers. Otherwise it's a good idea.

**Comment by bricewge** [comment]
This PR is related with the security vulnerability #31611. I don't think this should be merged as is until the issue is resolved.

File: nixos/modules/security/pam.nix [hunk 2]
```diff
@@ -287,7 +287,7 @@ let
           ${optionalString cfg.logFailures
               "auth required pam_tally.so"}
           ${optionalString (config.security.pam.enableSSHAgentAuth && cfg.sshAgentAuth)
-              "auth sufficient ${pkgs.pam_ssh_agent_auth}/libexec/pam_ssh_agent_auth.so file=~/.ssh/authorized_keys:~/.ssh/authorized_keys2:/etc/ssh/authorized_keys.d/%u"}
+              "auth sufficient ${pkgs.pam_ssh_agent_auth}/libexec/pam_ssh_agent_auth.so file=${concatStringsSep ":" config.services.openssh.authorizedKeysFiles}"}
```
Comment (fpletz) [inline nixos/modules/security/pam.nix]: This change will be obsolete by #62317.

**Review by fpletz — CHANGES_REQUESTED** [review]
:+1: on the change in principle.

I wonder if a special option like `users.mutableUserSshKeys` that defaults on the value of `users.mutableUsers` might make sense for people depending on the old behaviour with `users.mutableUsers` turned on. Or at least mention changing `services.openssh.authorizedKeysFiles` in the release notes, which we btw also need since this is a breaking change.

**Comment by SuperSandro2000** [comment]
@ip1981 please resolve the merge conflict.

**Pull request closed without merging**